### ISSUE TITLE
Setting the property loadWithKey to true within LoadConfig caused the Job to fail

### ISSUE THREAD

**Issue by MadhuMi** [CommBank/maestro#321]
When i  set the property loadWithKey to true within LoadConfig to test jobs that require the automatically generated row_id field as the last value , the job failed with following exception :-

cascading.pipe.OperatorException: [com.twitter.scalding.M...][com.twitter.scalding.RichPipe.eachTo(RichPipe.scala:478)] operator Each failed executing operation
	at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:107)
	at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:39)
	at cascading.flow.stream.SourceStage.map(SourceStage.java:102)
	at cascading.flow.stream.SourceStage.run(SourceStage.java:58)
	at cascading.flow.hadoop.FlowMapper.run(FlowMapper.java:130)
	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:417)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:332)
	at org.apache.hadoop.mapred.Child$4.run(Child.java:268)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:396)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1438)
	at org.apache.hadoop.mapred.Child.main(Child.java:262)
Caused by: java.util.NoSuchElementException: key not found: hdfs://nameservice1/dev/source/HLS/AMORTISATIONTRAN_SZGLAMR/2015/01/30/I5_OMN_HLS_SZGLAMR_20150130.DAT
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:58)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:58)
	at au.com.cba.omnia.maestro.core.task.GenerateKey.uid(LoadExecution.scala:263)
	at au.com.cba.omnia.maestro.core.task.GenerateKey.operate(LoadExecution.scala:297)
	at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:99)
	... 11 more

**Comment by SamRoberts** [comment]
@MadhuMi would you mind changing your execution slightly, so that before you call `load`, you first use `_ <- Execution.from(println(sources))` to print out the list of source files you are passing into the load function? And then show us what the list of source files looks like?

**Comment by MadhuMi** [comment]
@SamRoberts List(/dev/source/HLS/ADDRESS_SZCIMST/2014/11/21/I5_OMN_HLS_SZCIMST_COMM_ADDR_MRT900_20141121.DAT) . This is the list of source files that we provide to the load function.

**Comment by SamRoberts** [comment]
thanks, that is very helpful

**Comment by rowandavies** [comment]
When I add `generateKey = true` to the first test in `maestro.core.task.LoadExecutionSpec`, sources is `List(normal)`, and there appears to be a corresponding exception (the lines match up between the maestro versions for `maestro.core.task.GenerateKey`):
```
cascading.pipe.OperatorException: [com.twitter.scalding.M...][com.twitter.scalding.RichPipe.eachTo(RichPipe.scala:478)] operator Each failed executing operation
    at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:107)
    at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:39)
    at cascading.flow.stream.SourceStage.map(SourceStage.java:102)
    at cascading.flow.stream.SourceStage.run(SourceStage.java:58)
    at cascading.flow.hadoop.FlowMapper.run(FlowMapper.java:130)
    at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:417)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:332)
    at org.apache.hadoop.mapred.LocalJobRunner$Job$MapTaskRunnable.run(LocalJobRunner.java:268)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.NoSuchElementException: key not found: file:/tmp/hadoop/test-69cc3143-a8dd-4ac9-9374-ed16bdfb9d66/mapred/user/filtered/data.txt
    at scala.collection.MapLike$class.default(MapLike.scala:228)
    at scala.collection.AbstractMap.default(Map.scala:58)
    at scala.collection.MapLike$class.apply(MapLike.scala:141)
    at scala.collection.AbstractMap.apply(Map.scala:58)
    at au.com.cba.omnia.maestro.core.task.GenerateKey.uid(LoadExecution.scala:303)
    at au.com.cba.omnia.maestro.core.task.GenerateKey.operate(LoadExecution.scala:337)
    at cascading.flow.stream.FunctionEachStage.receive(FunctionEachStage.java:99)
    ... 12 more
```

**Comment by rowandavies** [comment]
The key that's not found is what's returned by `flow.getProperty(CASCADING_SOURCE_PATH).toString`, which seems to be a full path, unlike what's in `sources` in these two cases.

**Comment by SamRoberts** [comment]
Right, the cascading source path property is a fully qualified path, whereas none of the load inputs are. The right solution is probably for `load` to create a FileSystem instance from the execution Config and use it to fully qualify the source paths is is being passed.

Also, we need a unit test that tests the `loadWithKey` property.

**Comment by rowandavies** [comment]
Okay, I'm close to having a fix for this.  Resolving the sources seems to work for the original instance of the issue.  

Although, testing with `LoadExecutionSpec` still fails because `normal` resolves to something like `file:/tmp/hadoop/test-857a83c0-76a2-437c-b341-094caff5b41c/mapred/user/normal` which is different from `file:/tmp/hadoop/test-857a83c0-76a2-437c-b341-094caff5b41c/mapred/user/normal/data.txt`.

I'm unsure of where the file name `data.txt` is added, but manually changing the call to use `normal/data.txt` avoids the exception.  It still fails, I guess because I need to add to the thrift a field for the generated key.   Working on that now to create a proper test for the case when `generateKey` is true.

**Comment by rowandavies** [comment]
I've created a branch and pushed the path resolution fix:  
    https://github.com/CommBank/maestro/tree/generateKeyPathFix

This still needs an internal maestro test with an extended thrift type.

**Comment by SamRoberts** [comment]
@rowandavies Hadoop encourages users to pass around paths to the directory containing files, instead of paths to the file themselves. In that sense, the sources returned by `upload` are wrong, and the directories in the load spec are right. Hadoop handles both possibilities (either pass in the full file name, or it's immediate parent directory), so I think our load method should handle both possibilities too.

I guess there are a number of approaches you could take:
 - You could blow up the list of source files passed to `load` to include child files, so that you get the list of actual files, and calculate hashes for each file separately.

 - You could check to see if both the CASCADING_SOURCE_PATH and, if it is a file, it's immediate parent directory are in the map of hashes.

 - You could dynamically create the map of files to file hashes during the MR job, based solely on the CASCADING_SOURCE_PATH property of each file.

**Comment by MadhuMi** [comment]
I have tested the fix. the issue is found to be resolved

**Comment by rowandavies** [comment]
@SamRoberts I've implemented the first suggestion.  But I'm concerned that this affects the much more common generateKey=false case, so I'm still considering the other two.  
* The second one I'm unsure whether it matters if a directory contains multiple files which are all assigned the same hash.
* The third seems nicely robust, but seems to require a mutable `hashes` map.  Although, turning `hashes` instead into a memoizing function seems like a good solution, I think I'll try this.

**Comment by SamRoberts** [comment]
Right, the mutability should be an implementation detail of the hashes function, and should not affect it's referential transparency.

**Comment by SamRoberts** [comment]
Oh, just to add: I also prefer the third option, and share your concerns about the first and second one.

**Comment by rowandavies** [comment]
There is a tentative fix for this issue, but it's not completely satisfactory.
* It implements Sam's first suggestion above of blowing up the list of source files, and does this only when generateKey=true.   It ignores files starting with `_`, but this isn't included in the CDH5 branch (including the full path is a work around).
* This doesn't exactly match the CASCADING_SOURCE_PATH property for local sources unless they begin with `file:`, so the current work around is to always include this when generateKey=true.
* It's unclear how to turn this into a robust solution so that it always exactly matches the CASCADING_SOURCE_PATH property.

**Comment by stephanh** [comment]
What about taking the two different paths converting them to a HDFS `Path` and then use `makeQualified` to get the same representation?

**Comment by SamRoberts** [comment]
Rather then guessing in advance how CASCADING_SOURCE_PATH is going to work, I think the best approach is to try and get my third suggestion working. I think you can probably pass a custom serializable `Function` sub-class which uses a `Map` to memoise it's results.

**Comment by rowandavies** [comment]
@SamRoberts @stephanh Okay, I've been swapping between the two alternatives.   The second version so far has been awkward just because using the Hdfs monad and passing the filesystem object doesn't seem to fit so well with using it for each row, but it will work.   Currently I have the third version working - the root cause of the planner failure seems to be that `MessageDigest` instances aren't serializable.  Looking at the code now, I guess this is why it calls `MessageDigest.getInstance` for each row.  But, I think by making it lazy along with the memoizing hash function, the code can be simplified a lot.

**Comment by SamRoberts** [comment]
@rowandavies sounds good. Would you mind showing me what you have so far?

**Comment by rowandavies** [comment]
@SamRoberts Sure - in fact I forgot to say that I already committed what I have to the branch:
   generateKeyPathFixMemo

I haven't yet tested it with a real project.  And I think a lot of what's there in the generateKey code path can be simplified, and perhaps refactored so that `pipeWithDateAndKey` is a wrapper around `pipeWithDate`.

**Comment by rowandavies** [comment]
@SamRoberts Okay, now I understand how `prepare` and `operate` work in cascading, I've changed it to just calculate the hash of the filename in `prepare`, and removed the lookup table altogether.  This is much simpler, and should never have observably different performance.

I've committed a new version to the generateKeyPathFixMemo branch, but it doesn't really fit with the name of that branch, so I'll rebase it on master now, create a new branch and a PR.  Unless you see an issue with the new approach.
